src/plot_api/manage_arrays.js

```diff
@@ -0,0 +1,211 @@
+/**
+* Copyright 2012-2017, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+var nestedProperty = require('../lib/nested_property');
+var isPlainObject = require('../lib/is_plain_object');
+var noop = require('../lib/noop');
+var Loggers = require('../lib/loggers');
+var Registry = require('../registry');
+
+
+exports.containerArrayMatch = require('./container_array_match');
+
+var isAddVal = exports.isAddVal = function isAddVal(val) {
+    return val === 'add' || isPlainObject(val);
+};
+
+var isRemoveVal = exports.isRemoveVal = function isRemoveVal(val) {
+    return val === null || val === 'remove';
+};
+
+/*
+ * editContainerArray: for managing arrays of layout components in relayout
+ * handles them all with a consistent interface.
+ *
+ * Here are the supported actions -> relayout calls -> edits we get here
+ * (as prepared in _relayout):
+ *
+ * add an empty obj -> {'annotations[2]': 'add'} -> {2: {'': 'add'}}
+ * add a specific obj -> {'annotations[2]': {attrs}} -> {2: {'': {attrs}}}
+ * delete an obj -> {'annotations[2]': 'remove'} -> {2: {'': 'remove'}}
+ *               -> {'annotations[2]': null} -> {2: {'': null}}
+ * delete the whole array -> {'annotations': 'remove'} -> {'': {'': 'remove'}}
+ *                        -> {'annotations': null} -> {'': {'': null}}
+ * edit an object -> {'annotations[2].text': 'boo'} -> {2: {'text': 'boo'}}
+ *
+ * You can combine many edits to different objects. Objects are added and edited
+ * in ascending order, then removed in descending order.
```

Comment:
To me, they are two types of plotly.js errors:

- Invalid plot attribute combinations,
- Invalid input to `Plotly.___` methods.

I believe the former should never lead to errors as @rreusser said plotly.js should try to _do the best it can_. This is especially important for non-JS users that send `"data"` and `"layout"` and expect to get in return the _best possible_ :tm: graph.

This case :arrow_up: falls into the second category. `Plotly.__` method are called by JS users only* and throwing errors should help them debug. For example in `Plotly.plot(gd, fig)` if `gd` isn't right we [throw](https://github.com/plotly/plotly.js/blob/2533f04c12f0de69cc1b227d1eccd58b98c5d53f/src/plot_api/helpers.js#L25-L42) an error. I think the same should be expected for invalid relayout update object.

*: Note that updatemenus and slider now expose `Plotly.__` methods to non-JS users now. :expressionless: